design/design-doc-composition-validating-webhook.md

```diff
@@ -0,0 +1,391 @@
+   traversed by `patches` might not yet have been created when the Composition
+   is being validated.
+3. We do not have all the required CRDs locally, so we should get them from
+   the API server, if available.
+4. We do not have actual input values supplied from the Composite Resource that
+   are required to render valid and applicable Managed Resources.
+5. Some `alpha` features could require some additional interactions with
+   external systems, e.g. `CompositionFunctions`.
+
+### Solution Proposed
+
+In order to be able to have Compositions properly validated we should cover all
+the `resources` parts, while also taking into consideration all the constraints
+above.
+
+In order to address the possible absence of Managed Resources' CRDs, Composite
+Resources' CRDs or any other required external resource at validation time we
+should allow explicitly setting one of 2 modes, `strict` and `loose`(default) .
+These should be configurable through an annotation, e.g.
+`crossplane.io/composition-validation-mode`, on a Composition or maybe even
+globally via a flag, and would respectively imply that any missing resource
+should result in a direct rejection or a possibly incomplete validation.
+
+In `loose` mode we should emit warnings in case any of the preconditions for a
+check are not met, and skip further validation related to that check.
+
+Regarding the approach, I propose to implement a Composition Validating Webhook
+by validating each `resource` and final rendered `Composed
+Resource`, querying the API Server to get the needed schemas.
+
+I suggest we avoid validating `bases`. Doing that would require to ignore
+validating required fields, as those could be set at a later stage by patches
+or transformations. Assuming we implement full `patches` type coverage,
+validating `bases` per se can be avoided, as any error should either be caught
+validating the patches or can be attributed to the `base` itself.
+
+To validate `patches`, we will need to implement ad-hoc functions, because the
+current patching code is dealing directly with unstructured data and does not
+take into consideration schemas at all, therefore not reporting any schema
+related error. The code implemented for [upbound/upjet][upject-validation-code]
+or [crossplane-contrib/crossplane-lint][crossplane-lint-validation-code] could
+be good starting points for this.
+
+To render the Composed Resources, we will need to reuse the Composite resources
+reconciliation loop as much as possible, extracting the rendering logic to
+remove any write operation it currently performs. Which is similar to what has
+been done in the LSP [implementation][lsp-validation-code].
+
+All `patches` will have to be applied in order to have the final Managed
+Resource properly validated, so we'll need to provide all the required inputs
+to the reconciler, to properly apply them. Inputs could be either actual
+resources available or mocked ones.
```

Comment:
It sounds like we've decided to use mocked inputs. I think it would be useful to be explicit about that in this document, and explicit about what inputs we'll mock.